yarp/yarp_compiler.c

```diff
@@ -97,19 +373,140 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
           }
           return;
       }
+      case YP_NODE_ARRAY_NODE: {
+          yp_array_node_t *array_node = (yp_array_node_t *) node;
+          yp_node_list_t elements = array_node->elements;
+          if (elements.size == 1 && yp_static_node_literal_p(elements.nodes[0])) {
+              VALUE ary = rb_ary_hidden_new(1);
+              rb_ary_push(ary, yp_static_literal_value(elements.nodes[0]));
+              OBJ_FREEZE(ary);
+
+              ADD_INSN1(ret, &dummy_line_node, duparray, ary);
+          }
+          else {
+              for (size_t index = 0; index < elements.size; index++) {
+                  yp_compile_node(iseq, elements.nodes[index], ret, src, popped, compile_context);
+              }
+
+              if (!popped) {
+                  ADD_INSN1(ret, &dummy_line_node, newarray, INT2FIX(elements.size));
+              }
+          }
+
+          return;
+      }
       case YP_NODE_ASSOC_NODE: {
           yp_assoc_node_t *assoc_node = (yp_assoc_node_t *) node;
           yp_compile_node(iseq, assoc_node->key, ret, src, popped, compile_context);
           yp_compile_node(iseq, assoc_node->value, ret, src, popped, compile_context);
           return;
       }
+      case YP_NODE_ASSOC_SPLAT_NODE: {
+          yp_assoc_splat_node_t *assoc_splat_node = (yp_assoc_splat_node_t *)node;
+          yp_compile_node(iseq, assoc_splat_node->value, ret, src, popped, compile_context);
+
+          // TODO: Not sure this is accurate, look at FLUSH_CHUNK in the compiler
+          ADD_INSN1(ret, &dummy_line_node, newarraykwsplat, INT2FIX(0));
+
+          if (popped) {
+              ADD_INSN(ret, &dummy_line_node, pop);
+          }
+          return;
+      }
       case YP_NODE_BEGIN_NODE: {
           yp_begin_node_t *begin_node = (yp_begin_node_t *) node;
           if (begin_node->statements) {
               yp_compile_node(iseq, (yp_node_t *)begin_node->statements, ret, src, popped, compile_context);
           }
           return;
       }
+      case YP_NODE_CALL_NODE: {
+          yp_call_node_t *call_node = (yp_call_node_t *) node;
+
+          ID mid = parse_location_symbol(&call_node->message_loc);
```

Comment:
We should use `name` here, not `message_loc`. They're slightly different in some cases:

```ruby
foo.()          # name="call" message_loc=nil
foo.bar = 1     # name="bar=" message_loc="bar"
```